core-plugins/src/main/java/co/cask/hydrator/plugin/batch/file/FileCopyRecordWriter.java

```diff
@@ -0,0 +1,130 @@
+import org.apache.hadoop.fs.FSDataOutputStream;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.permission.FsPermission;
+import org.apache.hadoop.fs.s3a.S3AFileSystem;
+import org.apache.hadoop.fs.s3native.NativeS3FileSystem;
+import org.apache.hadoop.io.NullWritable;
+import org.apache.hadoop.mapreduce.RecordWriter;
+import org.apache.hadoop.mapreduce.TaskAttemptContext;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * The record writer that takes file metadata and streams data from source database
+ * to destination database
+ */
+public class FileCopyRecordWriter extends RecordWriter<NullWritable, AbstractFileMetadata> {
+  private FileSystem destFileSystem;
+  private String basePath;
+  private Boolean enableOverwrite;
+  private int bufferSize;
+  // buffer size defaults to 1 MB
+  private static final int defaultBufferSize = 2 << 20;
+  private static final Logger LOG = LoggerFactory.getLogger(FileCopyRecordWriter.class);
+
+  public FileCopyRecordWriter(Configuration conf)
+    throws Exception {
+    this.destFileSystem = FileSystem.get(conf);
+    this.basePath = conf.get(FileCopyOutputFormat.BASE_PATH);
+    this.enableOverwrite = conf.getBoolean(FileCopyOutputFormat.ENABLE_OVERWRITE, false);
+    this.bufferSize = conf.getInt(FileCopyOutputFormat.BUFFER_SIZE, defaultBufferSize);
+    this.sourceFilesystemMap = new HashMap<>();
+  }
+
+  // source filesystems
+  private Map<AbstractFileMetadata.Credentials, FileSystem> sourceFilesystemMap;
+
+  @Override
+  public void write(NullWritable key, AbstractFileMetadata fileMetadata) throws IOException, InterruptedException {
+    // get source database connection
+    AbstractFileMetadata.Credentials sourceCredentials = fileMetadata.getCredentials();
+    if (!sourceFilesystemMap.containsKey(sourceCredentials)) {
+      sourceFilesystemMap.put(sourceCredentials, getSourceFilesystemFromCredentials(sourceCredentials));
+    }
+    FileSystem sourceFilesystem = sourceFilesystemMap.get(sourceCredentials);
+
+    // construct file paths for source and destination
+    Path srcPath = new Path(fileMetadata.fullPath);
+    Path destPath = new Path(basePath + "/" + fileMetadata.basePath);
+    FsPermission permission = new FsPermission(fileMetadata.getPermission());
+
+    if (fileMetadata.isFolder) {
+      // create an empty folder and return
+      if (!destFileSystem.exists(destPath) && sourceFilesystem.isDirectory(srcPath)) {
+        destFileSystem.mkdirs(destPath, permission);
+      }
+      return;
+    } else if (!sourceFilesystem.exists(srcPath)) {
+      // file doesn't exist in source, return immediately
+      return;
+    }
+
+    // immediately return if we dont want to overwrite and file exists in destination
+    if (!enableOverwrite && destFileSystem.exists(destPath)) {
```

Comment:
I think it does some setup and cleanup, but you may need to make sure you're using whatever directories it is setting/cleaning up instead of the destination one specified by the plugin.